### PR TITLE
Introduce AssetOrArchive to support the SDK's Asset type which can be both

### LINKED ISSUE #237: Asset properties can be Archives at runtime
When defining a property of type `resource.Asset`, the generated schema will contain a `$ref` of `pulumi.json#/Asset`. However, in the SDKs, an "asset" is represented by `pulumi.AssetOrArchive` because an archive is currently always a valid asset.

Therefore, when deserializing the PropertyValue for the provider to use, it's valid for the user to have specified one of the Archive types which then isn't deserializable into the `resource.Asset` field within the provider.

1. We need instead a new `resource.AssetOrArchive` type to be used with the go-provider.
2. We need to handle reflecting over `resource.AssetOrArchive`
3. Using `resource.Asset` in a provider field should be considered an error. Though initially this should be a warning to avoid a breaking change but should be upgraded to an error on the next major release.

### PR DESCRIPTION
**Pull request by thomas11**
Resolves #237.

This PR addresses an issue with the design of the [assets and archives support](https://www.pulumi.com/docs/concepts/assets-archives/) in Pulumi core. Quoting #237:

> When defining a property of type resource.Asset, the generated schema will contain a $ref of pulumi.json#/Asset. However, in the SDKs, an "asset" is represented by pulumi.AssetOrArchive because an archive is currently always a valid asset.

> Therefore, when deserializing the PropertyValue for the provider to use, it's valid for the user to have specified one of the Archive types which then isn't deserializable into the resource.Asset field within the provider.

To work around this, we define a new type `types.AssetOrArchive` here in pulumi-go-provider. In the schema, it's represented as a standard `pulumi.json#/Asset` for interoperability. An incoming core `Asset` is translated into a `types.AssetOrArchive`, in `Decode`, with either Asset or Archive populated. Vice versa, `Encode` pulls the asset/archive out of AssetOrArchive.

## Notes
- The decode and encode implementations are bespoke and there might be opportunities to move it to a new, more generic extension point.

### CODE REVIEW

File: infer/internal/ende/ende.go
Comment (iwahbe) [inline infer/internal/ende/ende.go:199]: I'm amazed that this works, but if tests show that it does...

**Review by iwahbe — COMMENTED** [review]
The serde stuff looks good to me. This is the kind of place that I'd really appreciate integration tests to lock in good behavior.

File: infer/internal/ende/ende.go
Comment (thomas11) [reply infer/internal/ende/ende.go:199]: Could you elaborate on what you expected?

Comment (iwahbe) [reply infer/internal/ende/ende.go:199]: I'm not sure what I expected here. Probably an end run around pulumi/pulumi's mapper.

File: infer/types/asset.go
Comment (iwahbe) [inline infer/types/asset.go:21]: I thought we discussed using a special sig here to ensure we found `AssetOrArchive` and not some other type.
```suggestion
	Asset   *resource.Asset   `pulumi:"somereallylongsigforassets,optional"`
	Archive *resource.Archive `pulumi:"somereallylongsigforarchives,optional"`
```

File: infer/internal/ende/ende.go
Comment (iwahbe) [inline infer/internal/ende/ende.go:335]: Instead of calling it `"asset"`, lets use a sig:
```suggestion
				if rawAsset, ok := aMap["somereallylongsigforassets"]; ok {
```

This way we can be pretty sure that if we see the key, it's what we are looking for.

File: infer/resource.go
Comment (iwahbe) [inline infer/resource.go:1251]: Can we not adjust whitespace in files we don't otherwise touch?
```suggestion
```

File: infer/internal/ende/ende.go
Comment (iwahbe) [inline infer/internal/ende/ende.go:330]: Can we use the [constant in `pulumi/pulumi`](https://github.com/pulumi/pulumi/blob/5a83cfe6e4b77705240499134fae29d68c3c3ddf/sdk/go/common/resource/sig/sig.go#L18-L21)?
```suggestion
						if kind, ok := asset[sig.Key]; ok {
							if kind, ok := kind.(string); ok && kind == sig.AssetSig {

```

File: infer/internal/ende/ende.go
Comment (iwahbe) [inline infer/internal/ende/ende.go:325]: Can we make this a top level function to avoid the nesting?

File: infer/types/asset.go
Comment (iwahbe) [inline infer/types/asset.go:25]: Can we add a test that we will enforce a required `AssstOrArchive`? I want to make sure that this is safe:

```go
type MyArgs struct {
   Input types.AssetOrArchive `pulumi:"input"`
}

func (...) Create(...) ... {
    if args.Asset == nil && args.Archive == nil && !preview {
        panic("THIS SHOULD NEVER BE POSSIBLE")
    }
}
```

File: infer/internal/ende/ende.go
Comment (iwahbe) [inline infer/internal/ende/ende.go:329]: Let's make sure we error (or panic) if both the asset and archive keys are set.

File: infer/types/asset.go
Comment (iwahbe) [inline infer/types/asset.go:25]: This needs a doc comment explaining what it is and why you should use it. It *must* specify that this represents a union and setting both `Asset` and `Archive` will fail to serialize.

File: infer/resource.go
Comment (thomas11) [reply infer/resource.go:1251]: I had touched and apparently left this artifact when I reversed my changes.

File: infer/internal/ende/ende.go
Comment (danielrbradley) [inline infer/internal/ende/ende.go:376]: This line is missing test coverage

File: infer/internal/ende/ende.go
Comment (danielrbradley) [inline infer/internal/ende/ende.go:390]: These lines have no unit test coverage - probably worth adding.

File: infer/schema.go
Comment (danielrbradley) [inline infer/schema.go:135]: Should we add a warning here if they're using the previous Asset-only type above as it doesn't round-trip and they should migrate to AssetOrArchive?

**Review by danielrbradley — COMMENTED** [review]
This looks pretty close. I'm happy the implementation looks correct, just a few last bits of polish.

File: infer/internal/ende/ende.go
Comment (danielrbradley) [reply infer/internal/ende/ende.go:364]: Here's an example of a function we could extract to make this handling easier:

```go
func extractAssetOrArchive(aMap map[string]any, typeSig, kindSig string) (map[string]any, bool) {
	insideType, hasTypeSig := aMap[kindSig]
	if !hasTypeSig {
		return nil, false
	}
	insideTypeMap, ok := insideType.(map[string]any)
	if !ok {
		return nil, false
	}
	kind, hasKind := insideTypeMap[sig.Key]
	if !hasKind {
		return nil, false
	}
	kindStr, ok := kind.(string)
	if !ok {
		return nil, false
	}
	if kindStr != typeSig {
		return nil, false
	}
	return insideTypeMap, true
}
```

We can then be more specific on how we want to handle each error case too - panic or error etc.

Comment (danielrbradley) [reply infer/internal/ende/ende.go:364]: The method above then simplifies to:

```go
	if aMap, ok := a.(map[string]any); ok {
		if asAsset, isAsset := extractAssetOrArchive(aMap, sig.AssetSig, types.AssetSignature); isAsset {
			return resource.NewObjectProperty(resource.NewPropertyMapFromMap(asAsset)), true
		}
		if asArchive, isArchive := extractAssetOrArchive(aMap, sig.ArchiveSig, types.ArchiveSignature); isArchive {
			return resource.NewObjectProperty(resource.NewPropertyMapFromMap(asArchive)), true
		}
	}
````

File: infer/internal/ende/ende.go
Comment (iwahbe) [reply infer/internal/ende/ende.go:367]: I believe that this is when we are serializing. We need to serialize `AssetOrArchive` into either an asset or an archive, and we can't do that. We can't compress `struct { Asset, Archive }` into an `Asset` or `Archive` without losing information. It is *always* a provider bug and right now `panic`ing is a reasonable option. It would be better to return an error, but it's not clear how.

Comment (iwahbe) [reply infer/internal/ende/ende.go:367]: Platform will not create a new type that has our sigs (we assume that sigs won't conflict by accident).

File: infer/schema.go
Comment (thomas11) [reply infer/schema.go:151]: Screenshot 2024-06-11 at 20 39 46

File: infer/internal/ende/ende.go
Comment (thomas11) [reply infer/internal/ende/ende.go:371]: I don't quite see the simplification here TBH. I've commented and slightly reorganized the method in the latest commit, take a look if that looks better to you.

File: infer/internal/ende/ende.go
Comment (danielrbradley) [reply infer/internal/ende/ende.go]: It was the other way round - when we're parsing - we're adding a panic if we recieve something which contains both asset or archive information. My suggestion here would be to just check if it's an asset first, then check if it's an archive. The code below would already do that if the panic removed, if I read it correctly, though it prefers archive over asset.

File: infer/internal/ende/ende.go
Comment (danielrbradley) [reply infer/internal/ende/ende.go:371]: It's certainly partially subjective, so I wouldn't block on this.

Some observations of what makes it look complex to me are:
1. 5 levels of nested if statements means that the error handling paths are somewhat implicit rather than explicit - falling through to the null return (unless it ends up at the panic). The `extractAssetOrArchive` style would give the option for returning specific error messages for each case if we needed them too easier than adding `else` blocks for each `if` level. 
2. The panic "Encountered an unknown kind in an AssetOrArchive" is not adjacent to the condition that would trigger it - making the reader track back to work out the context. I.e. it's inverse of the branch it didn't take 4 lines above (`kind == sig.AssetSig || kind == sig.ArchiveSig`).
3. It's tricky follow if we're looking at an asset or we're looking at an archive, though it then appears it kinda doesn't matter after line 765 when we set the `raw` variable.

File: infer/internal/ende/ende.go
Comment (iwahbe) [reply infer/internal/ende/ende.go]: We should panic on ambiguity, as it is always a provider or framework bug. The user will never see the ambiguity and the engine can only pass ambiguous information if it copies the sigs we use, which it should not.

Comment (danielrbradley) [reply infer/internal/ende/ende.go]: I agree if it's user or framework error, but this data is coming from the engine – for which we don't have a super-good specification of what they might send, so I'd consider being defensive over making contract assertions. But perhaps it's reasonable here to fail hard because it would be a weird change engine side that _hopefully_ they wouldn't make because it's likely not backward compatible (not that we define backward compatibility in our protocols either 😞).

Comment (iwahbe) [reply infer/internal/ende/ende.go]: There is **no** remotely backwards compatible change (beyond copying the sigs we created so they wouldn't be copied, on purpose or by accident) that would allow this to be an engine bug. It will **always** be a provider author or framework bug, and it should panic to allow the responsible party to fix it, instead of doing the wrong thing for end users.

File: infer/internal/ende/ende.go
Comment (thomas11) [reply infer/internal/ende/ende.go:371]: The paths leading to the default return are not error paths. It's entirely expected that we encounter maps that are not AssetOrArchive, and fall through.

Comment (danielrbradley) [reply infer/internal/ende/ende.go:371]: There's also implicit error paths too:
- If it's got the asset type sig, but it's not a map within
- If it's an asset type sig, but has no kind field
- The kind field is not a string

These all just return null, which I think it fine, but it's implicit.

File: infer/schema.go
Comment (thomas11) [reply infer/schema.go:152]: Hmm, I just noticed that this also causes:
```
$ pulumi package get-schema /Users/tkappler/pulumi/pulumi-command/bin/pulumi-resource-command
warning: Please use pulumi-go-provider's types.AssetOrArchive type instead of resource.Asset or resource.Archive. ...
```

The reason is that types are serialized recursively into the schema, so first `AssetOrArchive`, then its inner `Asset` and `Archive` types.

File: infer/internal/ende/ende.go
Comment (iwahbe) [inline infer/internal/ende/ende.go:319]: This is not true, and we no longer have literal signautes.

File: infer/schema.go
Comment (iwahbe) [reply infer/schema.go:135]: Let's leave a warning as a follow up PR.

File: infer/types/asset.go
Comment (iwahbe) [inline infer/types/asset.go:20]: Let's try to add a docstring to each exposed part of the API.

```suggestion
const (
    // The signature for the asset part of [AssetOrArchive]
    AssetSignature = "a9e28acb8ab501f883219e7c9f624fb6"
    // The signature for the archive part of [AssetOrArchive]
    ArchiveSignature = "195f3948f6769324d4661e1e245f3a4d"
)
```

Even better, could we move these to `ende` and keep them private. I don't want them to show up in autocomplete since users should never need to use them.

File: infer/types/asset.go
Comment (iwahbe) [inline infer/types/asset.go:1]:
```suggestion
// Copyright 2024, Pulumi Corporation.
```

**Review by iwahbe — APPROVED** [review]
This looks good to me. There are a couple simple changes to make before merging, but it mostly looks good to go.

File: infer/schema.go
Comment (thomas11) [reply infer/schema.go:135]: https://github.com/pulumi/pulumi-go-provider/issues/243

File: infer/types/asset.go
Comment (thomas11) [reply infer/types/asset.go]: Ah, good call. In contrast to the type, the constants are only used in ende.

**Pull request merged**
